libs/mrwebrtc/src/interop/remote_audio_track_interop.cpp

```diff
@@ -1,72 +1,123 @@
+  }
+}
+
+MRS_API void* MRS_CALL
+mrsRemoteAudioTrackGetUserData(mrsRemoteAudioTrackHandle handle) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(handle)) {
+    return track->GetUserData();
+  }
+  return nullptr;
+}
+
+void MRS_CALL
+mrsRemoteAudioTrackRegisterFrameCallback(mrsRemoteAudioTrackHandle trackHandle,
+                                         mrsAudioFrameCallback callback,
+                                         void* user_data) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(trackHandle)) {
+    track->SetCallback(AudioFrameReadyCallback{callback, user_data});
+  }
+}
+
+mrsResult MRS_CALL
+mrsRemoteAudioTrackSetEnabled(mrsRemoteAudioTrackHandle track_handle,
+                              mrsBool enabled) noexcept {
+  auto track = static_cast<RemoteAudioTrack*>(track_handle);
+  if (!track) {
+    return Result::kInvalidParameter;
+  }
+  track->SetEnabled(enabled != mrsBool::kFalse);
+  return Result::kSuccess;
+}
+
+mrsBool MRS_CALL
+mrsRemoteAudioTrackIsEnabled(mrsRemoteAudioTrackHandle track_handle) noexcept {
+  auto track = static_cast<RemoteAudioTrack*>(track_handle);
+  if (!track) {
+    return mrsBool::kFalse;
+  }
+  return (track->IsEnabled() ? mrsBool::kTrue : mrsBool::kFalse);
+}
+
+void MRS_CALL
+mrsRemoteAudioTrackOutputToDevice(mrsRemoteAudioTrackHandle track_handle,
+    bool output) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(track_handle)) {
+      track->OutputToDevice(output);
+  }
+}
+
+mrsBool MRS_CALL mrsRemoteAudioTrackIsOutputToDevice(
+    mrsRemoteAudioTrackHandle track_handle) noexcept {
+  if (auto track = static_cast<const RemoteAudioTrack*>(track_handle)) {
+    return track->IsOutputToDevice() ? mrsBool::kTrue : mrsBool::kFalse;
+  }
+  return mrsBool::kFalse;
+}
+
+mrsResult MRS_CALL
+mrsRemoteAudioTrackCreateReadBuffer(mrsRemoteAudioTrackHandle track_handle,
+                              mrsAudioTrackReadBufferHandle* audioBufferOut) {
+  if (!audioBufferOut) {
+      return Result::kInvalidParameter;
+  }
+
+  *audioBufferOut = nullptr;
+  if (auto track = static_cast<RemoteAudioTrack*>(track_handle)) {
+    auto buffer = track->CreateReadBuffer();
+    *audioBufferOut = buffer.release();
+    return Result::kSuccess;
+  }
+  return Result::kInvalidNativeHandle;
+}
+
+mrsResult MRS_CALL
+mrsAudioTrackReadBufferRead(mrsAudioTrackReadBufferHandle buffer,
+                            int sample_rate,
+                            int num_channels,
+                            mrsAudioTrackReadBufferPadBehavior pad_behavior,
+                            float* data_out,
+                            int* data_len_in_out,
+                            mrsBool* has_overrun_out) {
+  if (!sample_rate || !num_channels ||
+      !IsValidAudioTrackBufferPadBehavior(pad_behavior) || !data_len_in_out ||
+      *data_len_in_out < 0 || (*data_len_in_out && !data_out) ||
+      !has_overrun_out)
+  {
+    return Result::kInvalidParameter;
```

Comment:
LOL, best of luck to the caller if they get this and don't have the PDBs! Even artificially splitting the cases and codes might help narrow down the cause.

Reply:
+1
I tend to write `RTC_LOG(LS_ERROR) << "..."` now for each parameter so users can self-diagnose.

Reply:
Fixed

Reply:
I hope `#__VA_ARGS__` works with `clang` on Android...

Reply:
Godbolt says it's fine on clang